docs/architecture/grpc-for-wcf-developers/migrating-duplex-services.md

```diff
@@ -1,11 +1,481 @@
 ---
 title: Migrate WCF duplex services to gRPC - gRPC for WCF Developers
-description: TO BE WRITTEN
+description: Migrating various forms of WCF duplex service to gRPC streaming services
 author: markrendle
 ms.date: 09/02/2019
 ---
 
 # Migrate WCF duplex services to gRPC
 
+Now the basic concepts are in place, this section will look at the more complicated *streaming* gRPC services.
+
+There are multiple ways to use Duplex services in WCF. Some services are initiated by the client and then stream data from the server; other "Full Duplex" services might involve more ongoing two-way communication like the classic "Calculator" example from the WCF documentation. This chapter will take two possible WCF "Stock Ticker" implementations and migrate them to gRPC: one using a Server streaming RPC, and one a Bi-directional Streaming RPC.
+
+## Server streaming RPC
+
+In the server streaming WCF example, *SimpleStockPriceTicker*, there is a Duplex service where the client initiates the connection with a list of stock symbols, and the server uses the *callback interface* to send updates as they become available. The client implements that interface and does whatever it does with the stream of real-time information.
+
+### The WCF solution
+
+The WCF solution is implemented as a self-hosted NetTCP server in a .NET 4.x console application.
+
+#### The ServiceContract
+
+```csharp
+[ServiceContract(SessionMode = SessionMode.Required, CallbackContract = typeof(ISimpleStockTickerCallback))]
+public interface ISimpleStockTickerService
+{
+    [OperationContract(IsOneWay = true)]
+    void Subscribe(string[] symbols);
+}
+```
+
+The service has a single method with no return type, because it will be using the callback interface `ISimpleStockTickerCallback` to send data to the client in real time.
+
+#### The callback interface
+
+```csharp
+[ServiceContract]
+public interface ISimpleStockTickerCallback
+{
+    [OperationContract(IsOneWay = true)]
+    void Update(string symbol, decimal price);
+}
+```
+
+The implementations of these interfaces can be found in the solution, along with faked external dependencies to provide test data.
+
+### gRPC streaming
+
+The gRPC way of handling real time data is different. A call from client to server can create a persistent stream, which can be monitored for messages arriving asynchronously. Despite the difference, streams can be a more intuitive way of dealing with this data, and are more relevant in modern programming with the emphasis on LINQ, Reactive Streams, functional programming and so on.
```

Comment:
Are you aware that WCF can handle IEnumerable\<T\> parameters and return types when using streamed mode transfer so that the iterator approach you are doing with gRPC is doable with WCF. The only difference is that WCF doesn't support IAsyncEnumerable as it didn't exist at the time. But that's not needed on the producer side as the yield operation effectively makes the producer async between MoveNext calls. On the client side, you would block consuming but this is something we could (and should) implement in our .NET Core client implementation.

Reply:
I don't think it matters so much for this book that WCF can do the IEnumerable thing, it's more that gRPC doesn't do the duplex thing so streams are the alternative approach.